src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1258,37 +1235,46 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
                         // to prevents a dependency on the thread pool (which would also trigger deadlocks).
                         // The WriteBuffer notifies this command when the first buffer flush occurs, so that the
                         // send functions can switch to the special async mode when needed.
-                        var sendTask = NonMultiplexingWriteWrapper(connector, async, finalCt);
+                        var sendTask = NonMultiplexingWriteWrapper(connector, async, writeCt);
 
-                        // The following is a hack. It raises an exception if one was thrown in the first phases
-                        // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
-                        // still happen later and aren't properly handled. See #1323.
-                        if (sendTask.IsFaulted)
-                            sendTask.GetAwaiter().GetResult();
+                        try
+                        {
+                            // The following is a hack. It raises an exception if one was thrown in the first phases
+                            // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
+                            // still happen later and aren't properly handled. See #1323.
+                            if (sendTask.IsFaulted)
+                                sendTask.GetAwaiter().GetResult();
+                        }
+                        catch (OperationCanceledException e)
+                        {
+                            // User requested the cancellation - breaking the connection, as WriteBuffer doesn't do it
+                            throw connector.Break(e);
+                        }
 
                         // TODO: DRY the following with multiplexing, but be careful with the cancellation registration...
                         var reader = connector.DataReader;
                         reader.Init(this, behavior, _statements, sendTask);
                         connector.CurrentReader = reader;
                         if (async)
-                            await reader.NextResultAsync(finalCt);
+                            await reader.NextResultAsync(readCt);
                         else
                             reader.NextResult();
                         return reader;
                     }
-                    catch
+                    catch (Exception e)
                     {
                         connector.CurrentReader = null;
                         conn.Connector?.EndUserAction();
+
+                        if (connector.UserCancellationRequested && !(e is OperationCanceledException))
```

Comment:
If I understand correctly, here you want to make sure that regardless of the actual exception type, you want to throw OperationCanceledException - is that right? If so, is this the behavior we want (not sure why)? For example, if the user requested cancellation and around the same time there was some I/O error, why shouldn't we raise the latter?

Reply:
Yeah, you're correct - we should handle here only PostgresException, the one about the cancelled query.